### ISSUE TITLE
Missing quadrature rule for ReactionDiffusionElementMatrixProvider

### ISSUE THREAD

**Issue by hiptmair** [craffael/lehrfempp#178]
Objects of type `ReactionDiffusionElementMatrixProvider` can be constructed by specifying a collection of quadrature rules:
~~~
ReactionDiffusionElementMatrixProvider(
      std::shared_ptr<const UniformScalarFESpace<SCALAR>> fe_space,
      DIFF_COEFF alpha, REACTION_COEFF gamma,
      quad_rule_collection_t qr_collection);
~~~
So far, if a quadrature rule for a particular element type is missing, the code terminates, which is not really desirable, because one may request assembly for triangular meshes only and in this case quadrature rules for quadrilaterals or segments are not really needed.

I propose that unspecified quadrature rules are substituted with default quadrature rules as they are set in an alternative version of the constructor already.

**Comment by hiptmair** [comment]
I implemented the proposed approach in the branch bugfix_uscalfe

**Comment by hiptmair** [comment]
I created a related pull request.

**Comment by craffael** [comment]
Yeah the current behavior is definitely not optimal. Also thanks for the bugfix. I see however a bit the problem, that with your proposed fix, "hard to find" bugs can emerge:
E.g. User A may choose to use a specific quadrature rule for triangles which differs from your proposed sensible default (e.g. for mass lumping) and for the moment he only uses triangular meshes. Everything Works for User A. But later on a different user B reuses the same code for hybrid meshes and gets unexpected results. It will take quite a bit of time to find out, that no quadrature rule is passed for quadrilaterals to the ReactionDiffusionElementMatrixProvider.
Therefore I would suggest, that we change it so that an assert fails whenever `ReactionDiffusionElementMatrixProvider` is called with an entity for which there is no quadrature rule... I can update your PR tonight to reflect this behavior...

**Comment by hiptmair** [comment]
What you propose is the best solution: catching a missing quadrature rule, that is, an uninitialized `PrecomputedScalarReferenceFiniteElement` in `Eval()`. 

I think, in this situation the code should not terminate, but an exception should be thrown so that the user can be told, e.g., "Only triangles are supported. Please provide another mesh." 

Well, there are several other places in the LehrFEM++ code where instant termination is too harsh a measure when a data inconsistency is detected. Rather suitable exceptions should be thrown.

**Comment by hiptmair** [comment]
Raffel, please go ahead an implement the changes you have suggested in the PR.

**Comment by craffael** [comment]
Okay, thanks for the feedback. I've implemented a corresponding fix in PR #180 . Can you please merge it into master if it's fine for you :)

**Comment by craffael** [comment]
Can be closed since PR #180 is merged.
